### PR TITLE
bpo-43502: Convert PyExceptionClass_Check from macro to static inline function

### PR DESCRIPTION
**Pull request by erlend-aasland**
https://bugs.python.org/issue43502

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. @vstinner

Should the NEWS entry include more info? For example _"the macro was problematic because it reused its arguments twice"_.

File: Include/pyerrors.h [hunk 1]
```diff
@@ -47,11 +47,7 @@ PyAPI_FUNC(void) PyException_SetCause(PyObject *, PyObject *);
 PyAPI_FUNC(PyObject *) PyException_GetContext(PyObject *);
 PyAPI_FUNC(void) PyException_SetContext(PyObject *, PyObject *);
 
-/* */
-
-#define PyExceptionClass_Check(x)                                       \
-    (PyType_Check((x)) &&                                               \
-     PyType_FastSubclass((PyTypeObject*)(x), Py_TPFLAGS_BASE_EXC_SUBCLASS))
+#define PyExceptionClass_Check(exc) _PyExceptionClass_Check(_PyObject_CAST(exc))
```
Comment (vstinner) [inline Include/pyerrors.h:54]: In the limited C API, you cannot use a function excluded from the limited C API. You must define _PyExceptionClass_Check() in the same header file, before the macro uses it.

**Review by vstinner — COMMENTED** [review]
If the function is not used in performance critical hot code, the macro can be converted to a regular function (not a static function).

File: Include/pyerrors.h (hunk 1 above)
Comment (erlend-aasland) [reply Include/pyerrors.h:54]: Thanks, noted.

Comment (erlend-aasland) [reply Include/pyerrors.h:54]: Moved "back" to the limited API in a6b67e65c308379f90ac28d5d541ddb1fc0e8f79

**Comment by erlend-aasland** [comment]
> If the function is not used in performance critical hot code, the macro can be converted to a regular function (not a static function).

My gut feeling is to convert `PyExceptionClass_Check` to a static inlined function, not a regular function. Let me know if you want it to be a regular C function.

File: Misc/NEWS.d/next/C API/2021-03-15-15-06-16.bpo-43502.G-uhre.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Convert ``PyExceptionClass_Check`` from macro to static inline function.
```
Comment (vstinner) [inline Misc/NEWS.d/next/C API/2021-03-15-15-06-16.bpo-43502.G-uhre.rst]: I'm not sure that it's worth it to document such change. The double evaluation issue exists in Python 2.7 and is likely older. If it's clearly a bugfix, it should be documented and backported. It doesn't seem to be the case here.

**Review by vstinner — COMMENTED** [review]
Include/ contains more than 450+ functions defined as macros. I don't want to have to review 450 PRs to change all macros.

Do you have an idea on how many macros have the double evaluation bug?

**Comment by erlend-aasland** [comment]
> Include/ contains more than 450+ functions defined as macros. I don't want to have to review 450 PRs to change all macros.
> 
> Do you have an idea on how many macros have the double evaluation bug?

I doubt all of them are _clearly problematic_. I'll see if I can create a list of worst cases and put it on the issue.

File: Misc/NEWS.d/next/C API/2021-03-15-15-06-16.bpo-43502.G-uhre.rst (hunk 2 above)
Comment (erlend-aasland) [reply Misc/NEWS.d/next/C API/2021-03-15-15-06-16.bpo-43502.G-uhre.rst]: Removed in fe5b0ff58033f96f1cc7a05598e371266dde0d1d

**Comment by erlend-aasland** [comment]
I'm closing this until there's some consensus regarding bpo-43502.

**Pull request closed without merging**